# Doggo Names - Data Science Tutorial

This is a one hour workshop where you can give data science a go!

We will be using the [Python programming language](https://www.python.org/about/gettingstarted/), and a Python "library" (set of tools) called [pandas](https://pandas.pydata.org/), to analyse some data about dog registrations in Pennsylvania. Let's see what interesting insights we can get from this!

Here are some resources for learning more about Python and pandas:
- [Simon Carryer's pandas tutorial that this workshop was based off](https://github.com/SimonCarryer/pandas_tutorial)
- [Codecademy](https://www.codecademy.com/search?query=python)
- [Checkio](https://checkio.org/)
- [Datacamp Intro to Python for Data Science](https://www.datacamp.com/courses/intro-to-python-for-data-science)

The first thing we have to do is "import" the libraries we're going to use. We're going to import the `pandas` library, and call it `pd` for short.

In [1]:
import pandas as pd

Now, let's use pandas (`pd` for short) to read our file of data. Pandas has a handy [read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function to do this. We can pass it the URL of the online data, and it'll read that file into a big table called a `DataFrame`.
It will then save it to a `variable` that we've decided to call `df`, short for `DataFrame`.

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/SimonCarryer/pandas_tutorial/master/data/dog_registrations.csv')

Let's take a peek at what's in the data, by calling the `head` function on our DataFrame. You call functions by using round brackets.

In [4]:
df.head()

LicenseType                Breed    Color DogName  \
0         Dog Individual Female  AM PIT BULL TERRIER  SPOTTED  BUTTER   
1         Dog Individual Female  AM PIT BULL TERRIER    BROWN   SABLE   
2  Dog Individual Neutered Male                MIXED        .     YIP   
3           Dog Individual Male    DOBERMAN PINSCHER      RED   SABER   
4  Dog Individual Spayed Female                MIXED    BLACK   DAISY   

   OwnerZip  ExpYear        ValidDate  
0     15001     2007   5/1/2007 15:15  
1     15001     2007   5/1/2007 15:15  
2     15001     2007  4/11/2007 15:14  
3     15003     2007   4/5/2007 15:00  
4     15003     2007  5/25/2007 12:15

## Counting & Filtering

In [23]:
df['Breed'].value_counts()

MIXED                  14000
LABRADOR RETRIEVER      8294
LAB MIX                 5575
GOLDEN RETRIEVER        3566
BEAGLE                  3239
                       ...  
GREYHOUND SPANISH          1
OTTERHOUND                 1
MIN BULL TERRIER           1
POLISH LOWLND SHEEP        1
CATALAN SHEEPDOG           1
Name: Breed, Length: 291, dtype: int64

291 breeds, and the top 5 are mixed breeds, labs, lab mixes, golden retrievers, and beagles.

What if we wanted to count the number of a specific breed? We can filter the `Breed` column for just the rows where the breed is equal to a certain breed. Let's take... the gigantic adorable [Newfoundland](https://www.akc.org/dog-breeds/newfoundland/) as an example.

First, we do this neat trick where we compare an entire column with a value - in this case, the `Breed` column of `df` compared against the string `NEWFOUNDLAND`. 

In [25]:
df['Breed'] == 'NEWFOUNDLAND'

0        False
1        False
2        False
3        False
4        False
         ...  
97457    False
97458    False
97459    False
97460    False
97461    False
Name: Breed, Length: 97462, dtype: bool

Then, we ask our pandas DataFrame to only give us back the rows where this comparison is true - in other words, only the rows where the `Breed` column has `NEWFOUNDLAND` in it.

In [27]:
df[df['Breed'] == 'NEWFOUNDLAND']

LicenseType         Breed        Color  \
1017            Dog Individual Male  NEWFOUNDLAND  WHITE/BLACK   
1151   Dog Individual Neutered Male  NEWFOUNDLAND  WHITE/BLACK   
1467            Dog Individual Male  NEWFOUNDLAND        BLACK   
3293          Dog Individual Female  NEWFOUNDLAND        BLACK   
3351   Dog Individual Spayed Female  NEWFOUNDLAND  WHITE/BLACK   
...                             ...           ...          ...   
95529  Dog Individual Spayed Female  NEWFOUNDLAND        BLACK   
95809           Dog Individual Male  NEWFOUNDLAND        BLACK   
96377  Dog Individual Neutered Male  NEWFOUNDLAND  BLACK/BROWN   
96719           Dog Individual Male  NEWFOUNDLAND        BLACK   
97424         Dog Individual Female  NEWFOUNDLAND        BLACK   

                 DogName  OwnerZip  ExpYear        ValidDate  
1017              SAWYER     15024     2007   8/2/2007 16:09  
1151          PANDA BEAR     15025     2007  2/13/2007 11:08  
1467                AMOS     15025     2007  5/16/2007 15:22  
3293              BERTHA     15044     2007   8/2/2007 11:29  
3351               CHLOE     15044     2007  2/28/2007 16:06  
...                  ...       ...      ...              ...  
95529              DAISY     15238     2009  12/9/2008 12:44  
95809                 BO     15239     2009  5/20/2009 14:35  
96377  MR BIG BIGGIE BOY     15239     2009  6/11/2009 13:13  
96719           NORMAN G     15241     2009   6/2/2009 16:17  
97424            HARRIET     16046     2009   4/27/2009 8:24  

[227 rows x 7 columns]

Can I just take a moment to point out that someone called their giant Newfoundland `Mr Big Biggie Boy`?

Okay, so back to counting up the number of Newfies in Pennsylvania. We could just look at the number of rows there, but we might want to print that answer out a bit more nicely. You can use the `count` function.

If we `count` on the entire DataFrame (with all the columns), it's going to give us a count of all the columns. We could just select a random column so it only returns one column:

In [34]:
df[df['Breed'] == 'NEWFOUNDLAND']['LicenseType'].count()

227

But this is kind of random. When coding, it's important to keep your code as clear as possible, so that when someone else comes along and reads it, they don't get confused.

If you just choose one of the columns at random for the purpose of counting up rows*, it might put them off track and they start wondering, "Huh, I wonder what license types have to do with counting the number of Newfoundlands?"

It would be clearer if we created a column that's purely used for counting. Let's add another column to this DataFrame called `row_count`, and just fill it that column with the number `1` in each row.

_*It also has some other potentially unexpected behaviour that's explained in the `Grouping` section of [this original tutorial](https://github.com/SimonCarryer/pandas_tutorial/blob/master/Part%20Two%20-%20Filtering%2C%20Sorting%2C%20and%20Grouping.ipynb#Grouping)), but that's getting a bit technical so don't worry too much about that._

In [35]:
df['row_count'] = 1

In [36]:
df.head()

LicenseType                Breed    Color DogName  \
0         Dog Individual Female  AM PIT BULL TERRIER  SPOTTED  BUTTER   
1         Dog Individual Female  AM PIT BULL TERRIER    BROWN   SABLE   
2  Dog Individual Neutered Male                MIXED        .     YIP   
3           Dog Individual Male    DOBERMAN PINSCHER      RED   SABER   
4  Dog Individual Spayed Female                MIXED    BLACK   DAISY   

   OwnerZip  ExpYear        ValidDate  row_count  
0     15001     2007   5/1/2007 15:15          1  
1     15001     2007   5/1/2007 15:15          1  
2     15001     2007  4/11/2007 15:14          1  
3     15003     2007   4/5/2007 15:00          1  
4     15003     2007  5/25/2007 12:15          1

Now we can explicitly select the `row_count` column when counting up the dog registrations with `NEWFOUNDLAND` as the breed.

In [38]:
df[df['Breed'] == 'NEWFOUNDLAND']['row_count'].count()

227

Make it even prettier? Save the count to a new variable (called something like `number_of_newfies` or whatever), and use the `print` function to include that number in a nicely `f`ormatted sentence!

In [41]:
number_of_newfies = df[df['Breed'] == 'NEWFOUNDLAND']['row_count'].count()
print(f'The number of newfies in Pennsylvania is {number_of_newfies}!')

The number of newfies in Pennsylvania is 227!


### Problem 1: Counting & Filtering

We've just learnt how to count and filter the rows of a pandas DataFrame, using the example of different dog breeds. Give it a go with colours.

- What is the most common dog colour?
- How many `SPOTTED` dogs are in this dataset?

In [61]:
df['Color'].value_counts()

BLACK                   15916
BROWN                   11920
WHITE                    7673
WHITE/BLACK              7489
BLACK/BROWN              7289
                        ...  
BLACK/MERLE                 5
BROWN/ORANGE                5
BLACK/ORANGE                5
BLACK/ORANGE/BRINDLE        4
WHITE/MERLE                 2
Name: Color, Length: 80, dtype: int64

Black is the most common dog colour.

In [63]:
num_spotted_dogs = df[df['Color'] == 'SPOTTED']['row_count'].count()
print(f'There are {num_spotted_dogs} spotted dogs in the dataset.')

There are 4637 spotted dogs in the dataset.


## Graphs

So counting things is handy and all, but how can we think about presenting this information in a fun, visually appealing way?

Python has a handy library called [seaborn](https://seaborn.pydata.org/)

Hey look! Data! Our `DataFrame` contains some information about dog registrations in Pennsylvania, including some things like the dog name and breed.

One of the most common things you'll want to do with a DataFrame is pull out a single row or column, maybe to do some calculation with it (in the case of numeric data) or maybe to count values (for categoricals). Let's look at how to access parts of a DataFrame.

## Accessing Columns

There are a few different ways of accessing columns in a DataFrame, but the easiest way (in my opinion) is to use the `__getitem__` function, which has a convenient shorthand notation using square brackets.

In [4]:
df['Breed'].head()

0    AM PIT BULL TERRIER
1    AM PIT BULL TERRIER
2                  MIXED
3      DOBERMAN PINSCHER
4                  MIXED
Name: Breed, dtype: object

You'll see that we're using the `head` function again to see just the first 5 rows of the DataFrame (if you want more than 5 rows, you can just pass the number you want in the brackets after the `head` like `.head(10)`.

It's useful to remember that the above code is shorthand for calling the secret `__getitem__` funtion - it's a quicker and clearer way of writing `df.__getitem__('Breed')`. Under the hood, they're doing the same thing, which is looking for a "Breed" attribute in the `df` object, and returning it. You can also access the attribute directly, with `df.Breed`, but this is a bit risky, because there are some "reserved" terms which will give you unexpected results.

You can also use a list of column names.

In [5]:
df[['Breed', 'ExpYear']].head()

Breed  ExpYear
0  AM PIT BULL TERRIER     2007
1  AM PIT BULL TERRIER     2007
2                MIXED     2007
3    DOBERMAN PINSCHER     2007
4                MIXED     2007

### Problem One

In the cell below:
* Show the values in the "Color" column.
* Show the values in the "LicenseType" and "Breed" columns.
* show the first 20 values in the "Breed" column.

## Accessing Rows

You can access rows in a DataFrame the same way you can access columns, but the syntax is a little different. The first and simplest way is to get data by its row number, using the `iloc` function, like this:

In [6]:
df.iloc[3]

LicenseType    Dog Individual Male
Breed            DOBERMAN PINSCHER
Color                          RED
DogName                      SABER
OwnerZip                     15003
ExpYear                       2007
ValidDate           4/5/2007 15:00
Name: 3, dtype: object

Just like for getting columns, this is a shorthand for the secret `__getitem__` function that belongs to the `iloc` object. It may seem a bit pedantic to keep going on about these secret functions, but it's useful to remember later on, when you're doing more complex stuff and the notation is confusing.

`iloc` can also get a range of rows, like this:

In [7]:
df.iloc[4:9]

LicenseType         Breed        Color  \
4                    Dog Individual Spayed Female         MIXED        BLACK   
5                    Dog Individual Neutered Male         MIXED      SPOTTED   
6                    Dog Individual Spayed Female   RAT TERRIER        MULTI   
7                           Dog Individual Female  GER SHEPHERD  BLACK/BROWN   
8  Dog Senior Citizen or Disability Spayed Female    POMERANIAN          TAN   

   DogName  OwnerZip  ExpYear        ValidDate  
4    DAISY     15003     2007  5/25/2007 12:15  
5  SCOOTER     15003     2007  6/19/2007 12:13  
6    TINKY     15003     2007  7/13/2007 13:35  
7    AMICA     15003     2007  2/27/2007 11:52  
8    TAFFY     15003     2007  3/12/2007 15:57

Notice how what is returned when we asked for one row looks quite different than what was returned when we asked for many? This is one of the annoying things in Pandas. Depending on if you ask for one row or many, it will return a different kind of object. The first one is a `Series`, and the second one is a "slice" of a `DataFrame`. This leads to all kinds of complex problems, which we'll get into more later.

One way you can get around this is by always using a list when you ask for specific rows. That ensures that even if there's only one item in your list, you'll still be returned a `DataFrame`, not a `Series`.

In [8]:
df.iloc[[3]]

LicenseType              Breed Color DogName  OwnerZip  ExpYear  \
3  Dog Individual Male  DOBERMAN PINSCHER   RED   SABER     15003     2007   

        ValidDate  
3  4/5/2007 15:00

There's one more way to access rows in DataFrames, and it's called `loc`. This may sound similar to `iloc`, but it's a bit different. We won't go into it right now though.

### Problem Two

In the cell below:
* Return the 18th row (as a DataFrame, not a Series)
* Return the 20th to 30th rows
* EXTRA FOR EXPERTS: Return the 4th, 7th, and 10th rows

## Operations on Columns

Now we know how to access data in rows or columns. Now let's do some things with that data!

Let's pull out a column of data into its own variable:

In [9]:
breeds = df['Breed']

`breeds` is a `Series` object. Each row and column in a Pandas DataFrame is a Series, and when you operate on a single row or column, you get access to the functions and attributes of the Series class, not the DataFrame class. That's often very useful, and sometimes a bit annoying.

Let's have a look at some useful Series functions:

In [10]:
breeds.unique()

array(['AM PIT BULL TERRIER', 'MIXED', 'DOBERMAN PINSCHER', 'RAT TERRIER',
       'GER SHEPHERD', 'POMERANIAN', 'BEAGLE', 'AM ESKIMO DOG',
       'COLLIE MIX', 'SIB HUSKY', 'BOXER', 'AUS SHEPHERD',
       'CHIHUAHUA MIX', 'BOXER MIX', 'SILKY TERRIER', 'SHIH TZU',
       'MIN PINSCHER', 'TERRIER', 'DACHSHUND', 'BICHON FRISE',
       'LABRADOR RETRIEVER', 'LAB MIX', 'PARSON RUSSELL TERR',
       'BRITTANY SPANIEL', 'POODLE TOY', 'POODLE MIN', 'BORD COLLIE MIX',
       'CHIHUAHUA', 'GOLDEN RETRIEVER', 'ENG SPRINGER SPANIE',
       'LHASA APSO', 'BEARDED COLLIE', 'LLEWELLIN SETTER', 'LABRADOODLE',
       'AKITA', 'TAG', 'ROTTWEILER', 'LHASA APSO MIX', 'COCKER SPANIEL',
       'GER SHEPHERD MIX', 'SIB HUSKY MIX', 'YORKSHIRE TERR MIX',
       'ENG BULLDOG', 'DACHSHUND MIX', 'YORKSHIRE TERRIER', 'POINTER',
       'COCKAPOO', 'ENG SETTER', 'SCHNAUZER STANDARD',
       'SHETLAND SHEEPDOG', 'KEESHOND', 'BASSET HOUND',
       'ITALIAN GREYHOUND', 'BOUVIER DES FLANDRE', 'PUG',
       'POODLE STAND

In [11]:
breeds.count()

97462

In [12]:
breeds.value_counts()

MIXED                    14000
LABRADOR RETRIEVER        8294
LAB MIX                   5575
GOLDEN RETRIEVER          3566
BEAGLE                    3239
GER SHEPHERD              2969
AM PIT BULL TERRIER       2969
SHIH TZU                  2249
BOXER                     2130
GER SHEPHERD MIX          2105
CHIHUAHUA                 1862
DACHSHUND                 1852
YORKSHIRE TERRIER         1700
BEAGLE MIX                1484
COCKER SPANIEL            1449
PUG                       1345
ROTTWEILER                1281
PARSON RUSSELL TERR       1225
BICHON FRISE              1154
SHETLAND SHEEPDOG         1103
MALTESE                   1045
SIB HUSKY                 1025
TERRIER MIX                956
POMERANIAN                 843
AM PITT BULL MIX           751
SCHNAUZER MIN              723
DOBERMAN PINSCHER          716
BOSTON TERRIER             715
COCKAPOO                   708
BORD COLLIE                670
                         ...  
W SHEPHERD /GER SHEPH        1
OTTERHOU

That last one looks super useful! It counts the rows for each unique value in the column. Let's do some more stuff with that!

In [13]:
breed_counts = breeds.value_counts()

Now we have another variable, called `breed_counts`. This is another Series object, but it's a little bit different than `breeds`, because its index is a list of dog breeds, not just row numbers. That changes how we can interact with it in a few ways, which we won't get into yet.

Instead, let's focus on something else. Dog breeds contains numeric data, which means we can access some more `Series` functions:

In [14]:
breed_counts.sum()

97462

In [15]:
breed_counts.mean()

334.9209621993127

In [16]:
breed_counts.median()

43.0

In [17]:
breed_counts.describe()

count      291.000000
mean       334.920962
std       1107.333765
min          1.000000
25%          6.000000
50%         43.000000
75%        247.500000
max      14000.000000
Name: Breed, dtype: float64

One more thing: Operations on `Series` objects are almost always carried out row-wise. That means that whatever you do to them, it does to each element in the Series. Observe:

In [18]:
breed_counts/breed_counts.sum()

MIXED                    0.143646
LABRADOR RETRIEVER       0.085100
LAB MIX                  0.057202
GOLDEN RETRIEVER         0.036589
BEAGLE                   0.033233
GER SHEPHERD             0.030463
AM PIT BULL TERRIER      0.030463
SHIH TZU                 0.023076
BOXER                    0.021855
GER SHEPHERD MIX         0.021598
CHIHUAHUA                0.019105
DACHSHUND                0.019002
YORKSHIRE TERRIER        0.017443
BEAGLE MIX               0.015226
COCKER SPANIEL           0.014867
PUG                      0.013800
ROTTWEILER               0.013144
PARSON RUSSELL TERR      0.012569
BICHON FRISE             0.011841
SHETLAND SHEEPDOG        0.011317
MALTESE                  0.010722
SIB HUSKY                0.010517
TERRIER MIX              0.009809
POMERANIAN               0.008650
AM PITT BULL MIX         0.007706
SCHNAUZER MIN            0.007418
DOBERMAN PINSCHER        0.007346
BOSTON TERRIER           0.007336
COCKAPOO                 0.007264
BORD COLLIE   

In [19]:
breeds + ' = GOOD DOG'

0        AM PIT BULL TERRIER = GOOD DOG
1        AM PIT BULL TERRIER = GOOD DOG
2                      MIXED = GOOD DOG
3          DOBERMAN PINSCHER = GOOD DOG
4                      MIXED = GOOD DOG
5                      MIXED = GOOD DOG
6                RAT TERRIER = GOOD DOG
7               GER SHEPHERD = GOOD DOG
8                 POMERANIAN = GOOD DOG
9                     BEAGLE = GOOD DOG
10                    BEAGLE = GOOD DOG
11             AM ESKIMO DOG = GOOD DOG
12                COLLIE MIX = GOOD DOG
13              GER SHEPHERD = GOOD DOG
14                 SIB HUSKY = GOOD DOG
15                     BOXER = GOOD DOG
16                     MIXED = GOOD DOG
17              AUS SHEPHERD = GOOD DOG
18             CHIHUAHUA MIX = GOOD DOG
19                 BOXER MIX = GOOD DOG
20             SILKY TERRIER = GOOD DOG
21                  SHIH TZU = GOOD DOG
22              MIN PINSCHER = GOOD DOG
23                   TERRIER = GOOD DOG
24               RAT TERRIER = GOOD DOG


In [20]:
breeds == 'MIXED'

0        False
1        False
2         True
3        False
4         True
5         True
6        False
7        False
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15       False
16        True
17       False
18       False
19       False
20       False
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
97432    False
97433    False
97434    False
97435     True
97436    False
97437    False
97438     True
97439    False
97440    False
97441    False
97442     True
97443     True
97444    False
97445    False
97446     True
97447     True
97448    False
97449    False
97450    False
97451    False
97452    False
97453    False
97454    False
97455    False
97456    False
97457     True
97458    False
97459    False
97460    False
97461    False
Name: Breed, Length: 97462, dtype: bool

This last one is important! When you compare a Series using `=`, `>`, `<` etc. it returns a Series object with `True` or `False` for each element in the Series. That's going to come in handy later.

### Problem Three

* What is the most common colour of dog?
* What is the average number of dogs per zip code?

## Creating New Columns

This might not be essential right now, but I promise you it will become useful later on. You can create new columns in a `DataFrame` just by assigning a value to a new column name.

In [21]:
df['row_count'] = 1

In [22]:
df.head()

LicenseType                Breed    Color DogName  \
0         Dog Individual Female  AM PIT BULL TERRIER  SPOTTED  BUTTER   
1         Dog Individual Female  AM PIT BULL TERRIER    BROWN   SABLE   
2  Dog Individual Neutered Male                MIXED        .     YIP   
3           Dog Individual Male    DOBERMAN PINSCHER      RED   SABER   
4  Dog Individual Spayed Female                MIXED    BLACK   DAISY   

   OwnerZip  ExpYear        ValidDate  row_count  
0     15001     2007   5/1/2007 15:15          1  
1     15001     2007   5/1/2007 15:15          1  
2     15001     2007  4/11/2007 15:14          1  
3     15003     2007   4/5/2007 15:00          1  
4     15003     2007  5/25/2007 12:15          1

A "row count" column can come in handy for a lot of more complex operations later on. It's not exactly essential, but it can make your code a lot easier to understand.

You can also create a column from any `Series` object that has the same index as your `DataFrame`.

In [27]:
df['lower_name'] = df['Breed'].str.lower() # Don't worry about what that ".str" means for now.
df['random_number'] = pd.Series(pd.np.random.randint(0, 10, size=len(df))) # Don't worry too much about what this is doing either

In [28]:
df.head()

LicenseType                Breed    Color DogName  \
0         Dog Individual Female  AM PIT BULL TERRIER  SPOTTED  BUTTER   
1         Dog Individual Female  AM PIT BULL TERRIER    BROWN   SABLE   
2  Dog Individual Neutered Male                MIXED        .     YIP   
3           Dog Individual Male    DOBERMAN PINSCHER      RED   SABER   
4  Dog Individual Spayed Female                MIXED    BLACK   DAISY   

   OwnerZip  ExpYear        ValidDate  row_count           lower_name  \
0     15001     2007   5/1/2007 15:15          1  am pit bull terrier   
1     15001     2007   5/1/2007 15:15          1  am pit bull terrier   
2     15001     2007  4/11/2007 15:14          1                mixed   
3     15003     2007   4/5/2007 15:00          1    doberman pinscher   
4     15003     2007  5/25/2007 12:15          1                mixed   

   random_number  
0              1  
1              3  
2              1  
3              4  
4              0

### Problem Four

* Make a column in `df` that indicates whether the dog's Breed is "MIXED".
* EXTRA FOR EXPERTS: Make a column in `df` that indicates whether the dog is male or female